File: fluentcore/src/main/java/com/mcreater/fxfluent/controls/FluentButton.java
package com.mcreater.fxfluent.controls;

import com.mcreater.fxfluent.controls.abstractions.*;
import com.mcreater.fxfluent.controls.skin.FluentButtonSkin;
import com.mcreater.fxfluent.controls.value.StateMap;
import com.mcreater.fxfluent.xaml.ResourceDict;
import com.mcreater.fxfluent.xaml.XAMLManager;
import com.mcreater.fxfluent.xaml.style.AppColorTheme;
import javafx.scene.Node;
import javafx.scene.control.Button;
import javafx.scene.control.Skin;
import javafx.scene.layout.CornerRadii;

import static com.mcreater.fxfluent.controls.value.ControlMaps.Button.*;

public class FluentButton extends Button implements CornerRadiusable, Backgroundable, Foregroundable, Borderable, Dictable, Themeable {
    private AppColorTheme theme = AppColorTheme.SYSTEM;
    public FluentButton() {
        this.init();
    }

    public FluentButton(String var1) {
        super(var1);
        this.init();
    }

    public FluentButton(String var1, Node var2) {
        super(var1, var2);
        this.init();
    }
    private void init() {
        getStyleClass().add("fluent-button");
    }
    public String getUserAgentStylesheet() {
        return FluentButton.class.getClassLoader().getResource("css/FluentButton.css").toString();
    }

    protected Skin<?> createDefaultSkin() {
        return new FluentButtonSkin(this);
    }

    public CornerRadii getCornerRadii() {
        return new CornerRadii(4);
    }

    public StateMap getBackgroundRemap() {
        return isDefaultButton() ? BG_ACCENT_KEY_MAP : BG_KEY_MAP;
    }

    public StateMap getForegroundRemap() {
        return isDefaultButton() ? FG_ACCENT_KEY_MAP : FG_KEY_MAP;
    }

    public StateMap getBorderRemap() {
        return isDefaultButton() ? BRD_BOTTOM_ACCENT_KEY_MAP : BRD_BOTTOM_KEY_MAP;
    }

    public ResourceDict getResourceDict() {
        return XAMLManager.getDict(theme);
    }

    public void onUpdateTheme(AppColorTheme theme) {
        this.theme = theme;
        ((FluentButtonSkin) getSkin()).implUpdate();
    }
}

File: fluentcore/src/main/java/com/mcreater/fxfluent/controls/FluentButton.kt
package com.mcreater.fxfluent.controls

import com.mcreater.fxfluent.controls.abstractions.*
import com.mcreater.fxfluent.controls.skin.FluentButtonSkin
import com.mcreater.fxfluent.controls.value.ControlMaps.Button.Companion.BG_ACCENT_KEY_MAP
import com.mcreater.fxfluent.controls.value.ControlMaps.Button.Companion.BG_KEY_MAP
import com.mcreater.fxfluent.controls.value.ControlMaps.Button.Companion.BRD_BOTTOM_ACCENT_KEY_MAP
import com.mcreater.fxfluent.controls.value.ControlMaps.Button.Companion.BRD_BOTTOM_KEY_MAP
import com.mcreater.fxfluent.controls.value.ControlMaps.Button.Companion.FG_ACCENT_KEY_MAP
import com.mcreater.fxfluent.controls.value.ControlMaps.Button.Companion.FG_KEY_MAP
import com.mcreater.fxfluent.controls.value.StateMap
import com.mcreater.fxfluent.xaml.ResourceDict
import com.mcreater.fxfluent.xaml.XamlManager.Companion.getDict
import com.mcreater.fxfluent.xaml.style.AppColorTheme
import javafx.scene.Node
import javafx.scene.control.Button
import javafx.scene.control.Skin
import javafx.scene.layout.CornerRadii


open class FluentButton : Button, CornerRadiusable, Backgroundable, Foregroundable, Borderable, Dictable,
    Themeable {
    private var theme: AppColorTheme? = AppColorTheme.SYSTEM

    constructor() {
        init()
    }

    constructor(var1: String?) : super(var1) {
        init()
    }

    constructor(var1: String?, var2: Node?) : super(var1, var2) {
        init()
    }

    private fun init() {
        styleClass.add("fluent-button")
    }

    override fun getUserAgentStylesheet(): String {
        return FluentButton::class.java.getClassLoader().getResource("css/FluentButton.css")?.toString() ?: ""
    }

    override fun createDefaultSkin(): Skin<*> {
        return FluentButtonSkin(this)
    }

    override val cornerRadii: CornerRadii
        get() = CornerRadii(4.0)
    override val backgroundRemap: StateMap
        get() = if (isDefaultButton) BG_ACCENT_KEY_MAP else BG_KEY_MAP
    override val foregroundRemap: StateMap
        get() = if (isDefaultButton) FG_ACCENT_KEY_MAP else FG_KEY_MAP
    override val borderRemap: StateMap
        get() = if (isDefaultButton) BRD_BOTTOM_ACCENT_KEY_MAP else BRD_BOTTOM_KEY_MAP
    override val resourceDict: ResourceDict
        get() = getDict(theme!!)

    override fun onUpdateTheme(theme: AppColorTheme?) {
        this.theme = theme
        (skin as SkinUpdatable).implUpdate()
    }
}

